Uploaded First Draft of Seminar 19 (#26)
alexfertel/reports: src/grupo_2/equipo_2_seminario_19/Readme.md

## Seminario 2: C#

> 

Reviewer: ```suggestion
No hace falta añadir los integrantes del equipo, ni la tabla de contenidos (ToC)
```

<span style="font-weight: bold;font-size: 2rem">Integrantes</span>

<ol>
    <li>
        Joel David Hernández Cruz
    </li>
    <li>
        Leandro González Montesino
    </li>
    <li>
        Wendy Díaz Ramírez
    </li>
</ol>

<span style="font-weight: bold">Tabla de Contenido</span>

- [Orientación](#Ejercicio)
- [Diseño por Contratos](#que-es-el-diseño-por-contratos?)
- [Decoradores en Python](#decoradores-en-python)
  - [Funciones como objetos de primera clase](#funciones-como-objetos-de-primera-clase)
  - [Funciones anidadas](#funciones-anidadas)
  - [Funciones que retornan funciones](#funciones-que-retornan-funciones)
- [Decoradores Simples](#decoradores-simples)
- [Decorando clases](#decorando-clases)
- [Implementación](#implementación)
  - [Librerias usadas](#Librerías-usadas)
- [Ejemplo](#Ejemplo)
- [Por Implementar](#Por-implementar)

### Ejercicio:

> Para este seminario debemos proponer e implementar un mecanismo de contratos al estilo de ``CodeContract`` en el lenguaje ***Python***. Con nuestra implementación deberíamos ser capaces de hacer algo como lo siguiente:

Reviewer: Tampoco hace falta la orden

La idea de ***Meyer***, aunque genial, se basa en una metáfora muy simple. Entre el que diseña una clase y el cliente de la misma se establece un "contrato" que define los deberes y derechos de ambos. Estos deberes y derechos se van a representar, en la terminología de ``Eiffel``, en forma de **pre-condiciones**, **pos-condiciones** e **invariantes**, tal y como se muestra en la siguiente tabla.

|                     |     Deberes u obligaciones      |     Derechos o beneficios       |
|---------------------|---------------------------------|---------------------------------|
|Cliente de la clase  |**Pre-condiciones**              |**Pos-condiciones e invariantes**|
|Diseñador de la clase|**Pos-condiciones e invariantes**|**Pre-condiciones**              |

-   ***Pre-condición***:

>     Es una cláusula lógica que **debe estar libre de efectos colaterales**, y que debe **cumplirse como requisito para ejecutar el método al que está asociada**. Un uso frecuente de las **pre-condiciones** es validar los parámetros de entrada de los métodos (algo que no puede satisfacerse solo con la declaración y tipado estático de los parámetros).

Reviewer: Mmmm, depende del concepto de validar. Yo pudiera argumentar que tener tipado estático elimina algunas validaciones que hicieran falta en lenguajes con tipado dinámico.

Author: No me queda claro que quieres decir con esto. Lo que hacemos referencia es a las validaciones que no se satisfacen solo con tipado. Por ejemplo la pre-condicion que `cnt >= 0` no basta con saber que `cnt` es de tipo `int`. 🤔

Reviewer: Que hay cosas que si puedes validar teniendo tipado estático, por ejemplo, `type(cnt) == int` chequearía que `cnt` sea entero, sin embargo, esto es algo que no tienes que validar en un lenguaje estáticamente tipado, en Python si lo tendrías que hacer: la precondición `cnt >= 0` falla si `cnt` no es entero. Por eso digo que la definición de validar depende, porque yo pudiera interpretar que estas diciendo: "El tipado estático de los parámetros no es suficiente para validar la entrada" y esto depende de qué validación vayas a hacer. Hay veces que si es suficiente :)

Author: ```suggestion
    Es una cláusula lógica que **debe estar libre de efectos colaterales**, y que debe **cumplirse como requisito para ejecutar el método al que está asociada**. Un uso frecuente de las **pre-condiciones** es validar aserciones en los parámetros de entrada a los métodos que no puede satisfacerse solo con la declaración y tipado estático de los mismo.
```
Que tal asi?

Reviewer: Forget it, déjalo así, esto es metatranca mía...no es importante

-   ***Pos-condición***:

>     Es una cláusula lógica que **debe cumplirse luego de la ejecución del método**, y funge como garantía de lo que éste hace. Por tanto, las **pos-condiciones** **son evaluadas justo después de ejecutado el método** y **antes de retornar al código que llama**.

Reviewer: Qué es funge?

Author: Eso fue sacado del libro de Katrib :). Lo puedo cambiar supongo.
```suggestion
    Es una cláusula lógica que **debe cumplirse luego de la ejecución del método**, y sirve como garantía de lo que éste hace. Por tanto, las **pos-condiciones** **son evaluadas justo después de ejecutado el método** y **antes de retornar al código que llama**.
```

-   ***Invariante***:
    Es una cláusula que **establece las condiciones bajo las cuales el estado de un objeto es "correcto"**. Por tanto, los invariantes **se validan luego de creados los objetos** y **después de ejecutar cualquier método**.

Para terminar este pequeña ***intro*** al **Diseño por Contratos**, un ejemplo de un código sencillo en ``Eiffel`` que, aunque quizás no se encuentre muy familiarizado con la sintaxis, es fácil captar la esencia del ejemplo.

![Eiffel_code_1](..\resources\Eiffel_code.PNG)

Para mas información sobre el Patrón de Diseño por Contratos le compartimos los siguientes enlaces:

<span>Diseño por Contratos: </span>
<span style="font-weight: bold; font-size: 1rem;">Aunque principalmente recomendamos el Libro del profesor Miguel Katrib y Jorge Luis de Armas: Diseño por Contratos en .NET 4.0</span>

- [Wikipedia](https://es.wikipedia.org/wiki/Dise%C3%B1o_por_contrato)
- [es.qwe.wiki](https://es.qwe.wiki/wiki/Design_by_contract)
- [Koalite](http://blog.koalite.com/2014/01/diseno-por-contrato/)

----

### Decoradores en Python

> Primero debemos aclarar algunas características del lenguaje `Python` y como funcionan sus funciones, que les serán útiles para entender decoradores.

Reviewer: Ese "como" lleva tilde

Reviewer: No hables refiriéndote al lector. "Primero aclaremos algunas características de `Python`, en particular el comportamiento de las funciones."

Author: 😃 
```suggestion
Primero aclaremos algunas características del lenguaje `Python`, en particular el comportamiento de sus funciones.
```

##### Funciones como Objetos de Primera Clase:

> Esto significa que las funciones en `Python` son objetos y como cualquier otro objeto puede ser pasadas y usadas como argumentos:

Reviewer: pueden*

Author: ```suggestion
Esto significa que las funciones en `Python` son objetos y como cualquier otro objeto pueden ser pasadas y usadas como argumentos:
```

```python
def say_hello(name):
    return f"Hello {name}"

def be_awesome(name):
    return f"Yo {name}, together we are the awesomest!"

def greet_bob(greeter_func):
    return greeter_func("Bob")
```

> En este ejemplo tenemos 3 funciones `say_hello()` y `be_awesome()` son funciones normales que esperan un parámetro `name` de tipo *string*. La función `gree_bob()` sin embargo espera una función como su argumento. Es fácil saber cual sera la salida para el siguiente ejemplo:

Reviewer: `greet_bob`*. Además fijense que la función es el objeto sin paréntesis, es decir, `greet_bob`, no `greet_bob`, lo segundo se refiere al resultado de llamar a la funcion, en general deberían quitarlos dondequiera que lo usen.

```python
>>> greet_bob(say_hello)
'Hello Bob'

>>> greet_bob(be_awesome)
'Yo Bob, together we are the awesomest!'
```

> Importante notar aquí como las funciones `say_hello()` y `be_awesome()` son llamadas sin los paréntesis, esto significa que solo una referencia a la función es pasada como parámetro. 

Reviewer: Me parece que no tienen claro bien lo que significa que las funciones sean objetos. `greet_bob(be_awesome)` está llamando a la función `greet_bob` con `be_awesome` como argumentos, *en ningun momento se "llama" a `be_awesome`*. Lo que significa es simplemente lo que significaría en cualquier lenguaje que ud conocen: *se le pasa una **copia** de la referencia al objeto `be_awesome`*. Arreglen este pedazo en base a lo anterior.

Author: A ver eso nos queda claro, `greet_bob` recibe una copia de la referencia a esa funcion como argumento (lo cual recalcamos mas abajo) 
**Importante notar aquí como las funciones `say_hello()` y `be_awesome()` son usadas sin los paréntesis, esto significa que solo una copia de la referencia a esa función es pasada como parámetro.,** 

Asi mejor?

Reviewer: Si, lo único que yo quitaría los paréntesis :)

Author: ```suggestion
Importante notar aquí como las funciones `say_hello` y `be_awesome` son usadas sin los paréntesis, esto significa que solo una copia de la referencia a esa función es pasada como parámetro.
```

##### Funciones anidadas

Es posible definir funciones dentro de otras funciones:

```python
def parent():
    print("Printing from the parent() function")

    def first_child():
        print("Printing from the first_child() function")

    def second_child():
        print("Printing from the second_child() function")

    second_child()
    first_child()
```

La salida de llamar a la función `parent()` es justo lo que se espera.

```python
>>> parent()
'Printing from the parent() function'
'Printing from the second_child() function'
'Printing from the first_child() function'
```

##### Funciones que retornan funciones

`Python` también permite usar funciones como valor de retorno. Usando el ejemplo anterior:

```python
def parent(num):
    def first_child():
        return "Hi, I am Joel"

    def second_child():
        return "Call me David"

    if num == 1:
        return first_child
    else:
        return second_child
```

> Notar que estamos retornado una referencia a las funciones, así que la siguiente salida es lo esperado.

Reviewer: retornando*. No escribiría esto así referirse al comentario anterior.

Author: ```suggestion
Notar que estamos retornando una copia de la referencia a esas funciones, así que la siguiente salida es lo esperado.
```

Reviewer: retornando*

```python
>>> first = parent(1)
>>> second = parent(2)

>>> first
<function parent.<locals>.first_child at 0x7f599f1e2e18>

>>> second
<function parent.<locals>.second_child at 0x7f599dad5268>
```

#### Decoradores Simples

> Ahora que ya hemos visto algunas características de las funciones en `Python` veamos que son los decoradores.

Reviewer: qué*

Author: ```suggestion
Ahora que ya hemos visto algunas características de las funciones en `Python` veamos qué son los decoradores.
```

Reviewer: jajaja, papa es el otro que....

###### my_decorator

```python
def my_decorator(func):
    def wrapper():
        print("Something is happening before the function is called.")
        func()
        print("Something is happening after the function is called.")
    return wrapper
```

```python
def say_whee():
    print("Whee!")

say_whee = my_decorator(say_whee)
```

Esto no es nada nuevo, es simplemente lo que ya se vio en ejemplos anteriores, así que es fácil saber cual sera la salida de llamar a `say_whee()`.

```python
>>> say_whee()
'Something is happening before the function is called.'
'Whee!'
'Something is happening after the function is called.'
```

> En este ejemplo el decorador ocurre en la linea `say_whee = my_decorator(say_whee)`, la variable `say_whee` ahora apunta al `wrapper` que se define dentro de `my_decorator`. Pero `wrapper` tiene una referencia a la función original `say_whee` y es la que ejecuta entre los dos `prints`.

Reviewer: "el decorador ocurre"? Tal vez "el decorado ocurre" o  "la decoración ocurre". "apunta", "`wrapper` tiene una referencia", "es la que ejecuta (la referencia)" creo que no son la mejor forma de explicar lo que pasa.

Author: ```suggestion
En este ejemplo la acción de decorar ocurre en `say_whee = my_decorator(say_whee)`, `my_decorator()` es una función que retorna una copia de la referencia a la función `wapper` que se define dentro del mismo, luego la variable `say_whee` tiene la copia de esta referencia. Y a su vez  `wrapper` tiene una copia de la referencia a la función original `say_whee` y es esta la que ejecuta entre los dos `prints`.
```

Reviewer: Jajaja, que enredo....

> En simples términos:  ***decorators wrap a function, modifying its behavior.***

Reviewer: Nunca usen frases en inglés por gusto, cuál es el objetivo de poner esta línea, justo debajo la escriben en español.

Un decorador envuelve una función, agregándole o modificando su funcionalidad.

`Python` te permite un azúcar sintáctico para usar decoradores en el código de una manera mas fácil y legible. **Usando el símbolo @**, el anterior código quedaría entonces:

```python
@my_decorator
def say_whee():
    print("Whee!")

# Output: say_whee()
'Something is happening before the function is called.'
'Whee!'
'Something is happening after the function is called.'
```

Que pasaría entonces si quisiéramos decorar una función que reciba un argumento, usando [my_decorator](#my_decorator)

```python
@my_decorator
def greet(name):
    print(f"Hello {name}")
    
# Output
>>> greet("World")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: wrapper() takes 0 positional arguments but 1 was given
```

> Recordemos que nuestro decorador [my_decorator](#my_decorator) retorna una referencia a `wrapper` y esta es una función que no recibe ningún argumento. Esto se podría arreglar simplemente agregándole un parámetro a `wrapper` pero entonces la primera función `say_whee` dejaría de funcionar, ademas funciones de más de 1 parámetro seguirían fallando. La solución para este caso es usar [*args y **kwargs](https://stackoverflow.com/questions/36901/what-does-double-star-asterisk-and-star-asterisk-do-for-parameters) en la función `wrapper` así esta aceptaría un numero arbitrario de parámetros.

Reviewer: además* número*. "...arbitrario de *argumentos*".

Esta azúcar sintáctica que permite `Python` se traduce a lo siguiente.

```python
@my_decorator
def func():
    print("func")

@my_decorator
def func_with_args(a, b, c)
	print(f"func_with_args: {a}, {b}, {c}")

# 1. func() => my_decorator(func)()
# 2. func_with_args(1, 2, 3) => my_decorator(func)(1, 2, 3)
```

Por supuesto las declaraciones para decorar una función pueden ser 1 o muchas, veamos un ejemplo de esto:

```python
@my_decorator
@my_decorator
def func_with_args(a, b, c)
	pass

# Output
# En este caso al llamar a func_with_args(1, 2, 3) lo que se imprime es

# Something is happening before the function is called.
# Something is happening before the function is called.
# func_with_args: 1, 2, 3
# Something is happening after the function is called.
# Something is happening after the function is called.
```

----

#### Decorando Clases

> Además de decorar funciones también podemos decorar clases. Las clases pueden ser decoradas de dos maneras. Primero por supuesto puede simplemente decorar los métodos de la clase que quieras, en `Python` los mas comunes que incluso vienen `built-in` son `@classmethod`, `@staticmethod`, `@property`. Aquí les dejamos un ejemplo de uso de estos métodos y documentación sobre ellos.

Reviewer: más*

        """Factory method creating a circle with radius 1"""
        return cls(1)

    @staticmethod
    def pi():
        """Value of π, could use math.pi instead though"""
        return 3.1415926535
```

En esta clase:

-	`.cylinder_volume()` es un método regular de la clase.
-	`.radius` es una propiedad mutable. Como se aprecia en el ejemplo se le define un `setter`, por lo cual su valor puede ser alterado. **Es una propiedad** así que debe ser accedida sin paréntesis.
-	`.area` es una propiedad inmutable, al contrario que le ejemplo anterior no tiene un `setter` definido, así que su valor no puede ser cambiado.
-	`.unit_circle()` es un método de clase. No esta ligado a una instancia en particular de `Circle`. Estos métodos generalmente son usados como métodos de *fabrica*, que pueden crear instancias especificas de una clase.
-	`.pi()` un método estático. No depende de la clase `Circle`, excepto por el hecho de estar en su ***namespace***. Estos pueden ser llamados desde una instancia o desde la clase.

----

La otra forma de decorar clases, es decorando la clase completa.

```python
from decorators import my_decorator

@my_decorator
class Circle:
    pass
```

> Decorar una clase, no es muy distinto que decorar una función. Por supuesto la diferencia esta en que el decorador en lugar de recibir una función como argumento, recibe una clase. El decorador que vimos anteriormente [my_decorator](#my_decorator) funcionara igual si se le pasa una clase, lo que es probable que no haga lo que teníamos pensado cuando lo hicimos. Para mas información sobre decoradores les compartimos los siguientes enlaces.

Reviewer: Esa coma al principio sobra. está* funcionará* más*

Documentación sobre <span style="font-weight: bold">Python Decorators</span>.

- [Python-Course Site](https://www.python-course.eu/python3_decorators.php)
- [DataCamp Site](https://www.datacamp.com/community/tutorials/decorators-python)
- [GeeksforGeeks](https://www.geeksforgeeks.org/decorators-in-python/)
- [TowardsDataScience](https://towardsdatascience.com/decorators-in-python-9cf8ba95e8e7)

----

#### Implementación

> ##### Librerías usadas:

Reviewer: Bibliotecas

Reviewer: Librerias con tilde?

                try:
                    x = super(NewCls, self).__getattribute__(s)
                except AttributeError:
                    pass
                else:
                    return x
                x = self.oInstance.__getattribute__(s)
                if type(x) == type(self.__init__):  # it is an instance method
                    check_invariant(self, inva)
                    
                    return x
                else:
                    return x
        return NewCls
    return invariant
```

-----

##### Ejemplo:

```python
@contract_invariant(inva=lambda balance: balance > 0)
class Wallet:

    def __init__(self, balance):
        self.balance = balance

    @contract(require=lambda amount: amount >= 0, ensure=lambda balance: balance >= 0)
    def deposit(self, amount):
        self.balance += amount

    @contract(require=lambda amount, balance: amount >= 0 and amount <= balance, 					  ensure=lambda balance: balance >= 0)
    def withdraw(self, amount):
        self.balance -= amount

    @property
    def check_balance(self):
        print(f"You're current balance is: {self.balance}")
```

-----

##### Por implementar:

> En nuestra solución no se encuentra implementado el ***feature*** de Diseño por Contratos donde se puede tener acceso al valor de una variable antes de la ejecución del método, analicemos primero que para implementar este ***feature*** debemos salvar todas las variables en el ***scope (Dígase en sus parámetros definidos y en la instancia de la cual se llama)*** del método antes de su ejecución, ya que no sabemos cual se usará en el cuerpo de la expresión ***lambda***, esto no es mucho problema, el problema esta en que estas variables nuevas no podemos pasarlas como parámetro a la expresión ***lambda***, así que debemos añadirlas al ***scope*** global para que puedan ser usada y reescribirlas para los próximos usos, además se debe seguir algún tipo de convención de nombres a la hora de definir el contrato, algo como llamar a las variables con un prefijo `old_`, así si se tiene una variable `balance` entonces para referirnos al valor antes de la ejecución del método debería ser como `old_balance`.

Reviewer: Me parece que no es tan complicado, pero tienen razón en que no es objetivo del seminario. Faltan varias tildes en este párrafo. En general el seminario está muy bien, lo que tienen detalles de redacción, por favor, arréglenlos con detenimiento.
